CCHMapClusterController/CCHMapClusterOperation.m

```diff
@@ -97,20 +101,237 @@ + (MKMapRect)gridMapRectForMapRect:(MKMapRect)mapRect withCellMapSize:(double)ce
+// 3. Add all annotations that are within a certain distance to the cluster.
+// 4. Move any annotations out of an existing cluster if they are closer to another cluster.
+// 5. Remove those items from the list of candidate clusters.
+// 6. Add/remove items on the actual map as in the grid algorithm.
+//
+- (void) start_distance
+{
+	self.executing = YES;
+	
+	BOOL respondsToSelector = [_clusterControllerDelegate respondsToSelector:@selector(mapClusterController:willReuseMapClusterAnnotation:)];
+
+	double zoomLevel = CCHMapClusterControllerZoomLevelForRegion(self.mapViewRegion.center.longitude, self.mapViewRegion.span.longitudeDelta, self.mapViewWidth);
+	BOOL disableClustering = (zoomLevel > self.maxZoomLevelForClustering);
+
+	// Zoom scale * MK distance = screen points
+	MKZoomScale zoomScale = [self currentZoomScale];
+	
+	// The width and height of the square around a point that we'll consider later
+	double zoomSpecificSpan = _cellSize / zoomScale;
+	
+	// Annotations we've already looked at for a starting point for a cluster
+	NSMutableSet *visitedCandidates = [[NSMutableSet alloc] init];
+
+	// The MKAnnotations (single POIs and clusters alike) we want on display
+	NSMutableSet *results = [[NSMutableSet alloc] init];
+	
+	// Map a single POI MKAnnotation to its distance (NSNumber*) from its cluster (if added to one yet)
+	NSMapTable *distanceToCluster = [NSMapTable strongToStrongObjectsMapTable];
+	
+	// Map a single POI MKAnnotation to its cluster annotation (if added to one yet)
+	NSMapTable *itemToCluster = [NSMapTable strongToStrongObjectsMapTable];
+
+	// Existing annotations to re-use
+	NSMutableSet *annotationsToReuse = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotationsInMapRect:self.mapView.visibleMapRect]];
+	//NSMutableSet *annotationsToReuse = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotations]];
+	
+	// The CCHMapClusterAnnotations that are being reused
+	NSMutableSet *reusedAnnotations = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotations]];
+	
+	@synchronized(_allAnnotationsMapTree) {
+		
+		// Iterate over all POIs we know about
+		for (id<MKAnnotation> candidate in _allAnnotationsMapTree.annotations) {
+			
+			if ([visitedCandidates containsObject:candidate]) {
+				// Candidate is already part of another cluster.
+				continue;
+			}
+				
+			// searchBounds is a rectangle of size zoomSpecificSpan (map x and y,
+			// not LatLng), centered on the candidate item's point
+			MKMapPoint point = MKMapPointForCoordinate(candidate.coordinate);
+			MKMapRect searchBounds = [self.class createBoundsFromSpan:point span:zoomSpecificSpan];
+			
+			// Get list of MKAnnotations in that bounds
+			NSSet *clusterItems = [_allAnnotationsMapTree annotationsInMapRect:searchBounds];
+			
+			CCHMapClusterAnnotation *cluster;
+
+			if (_reuseExistingClusterAnnotations) {
+				// Check if an existing cluster annotation can be reused
```

Comment:
I'm not quite sure if this works as expected. I'd start simple and first make the algorithm work without reusing annotations. Then in the next iteration we can check how to adapt the old re-use logic to the new algo